src/backend/distributed/planner/function_call_delegation.c

```diff
@@ -449,3 +542,190 @@ ShardPlacementForFunctionColocatedWithReferenceTable(CitusTableCacheEntry *cache
 
 	return (ShardPlacement *) linitial(placementList);
 }
+
+
+/*
+ * Checks to see if the procedure is being executed on a worker after delegated
+ * by the coordinator. If the flag forcePushdown is set, capture the distribution
+ * argument value, to be used by the planner to make sure that function uses only
+ * the colocated shards of the distribution argument.
+ */
+void
+CheckDelegatedFunctionExecution(DistributedPlanningContext *planContext)
+{
+	List *targetList = planContext->query->targetList;
+	FromExpr *joinTree = planContext->query->jointree;
+
+	if (list_length(targetList) > 1 || list_length(joinTree->fromlist) > 1)
+	{
+		/* multiple target list or from list items */
+		return;
+	}
+
+	/*
+	 * Look for a function in the SELECT clause.
+	 */
+	FuncExpr *targetFuncExpr = NULL;
+	if (list_length(targetList) == 1)
+	{
+		TargetEntry *targetEntry = (TargetEntry *) linitial(targetList);
+		if (IsA(targetEntry->expr, FuncExpr))
+		{
+			targetFuncExpr = (FuncExpr *) targetEntry->expr;
+		}
+	}
+
+	/*
+	 * Look for a function in the FROM clause.
+	 */
+	FuncExpr *fromFuncExpr = FunctionInFromClause(joinTree->fromlist, planContext->query);
+
+	/*
+	 * Look for one of:
+	 * SELECT fn(...);
+	 * SELECT ... FROM fn(...);
+	 */
+	FuncExpr *funcExpr = NULL;
+	if (targetFuncExpr != NULL)
+	{
+		if (fromFuncExpr != NULL)
+		{
+			/* query is of the form: SELECT fn() FROM fn() */
+			return;
+		}
+
+		/* query is of the form: SELECT fn(); */
+		funcExpr = targetFuncExpr;
+	}
+	else if (fromFuncExpr != NULL)
+	{
+		/* query is of the form: SELECT ... FROM fn(); */
+		funcExpr = fromFuncExpr;
+	}
+	else
+	{
+		/* query does not have a function call in SELECT or FROM */
+		return;
+	}
+
+	DistObjectCacheEntry *procedure = LookupDistObjectCacheEntry(ProcedureRelationId,
+																 funcExpr->funcid, 0);
+	Assert(procedure != NULL);
+
+	if (!procedure->isDistributed || !procedure->forcePushdown)
+	{
+		/* not a delegated function call within a 2PC */
+		return;
+	}
+
+	/*
+	 * On the coordinator PartiallyEvaluateExpression() descends into an
+	 * expression tree to evaluate expressions that can be resolved to a
+	 * constant. Expressions containing a Var are skipped, since the value
+	 * of the Var is not known on the coordinator.
+	 */
+	Node *partitionValueNode = (Node *) list_nth(funcExpr->args,
+												 procedure->distributionArgIndex);
+	Assert(partitionValueNode);
+	partitionValueNode = strip_implicit_coercions(partitionValueNode);
+
+	if (IsA(partitionValueNode, Param))
+	{
+		Param *partitionParam = (Param *) partitionValueNode;
+
+		if (partitionParam->paramkind == PARAM_EXTERN)
+		{
+			/* Don't log a message, we should end up here again without a parameter */
+			DissuadePlannerFromUsingPlan(planContext->plan);
+			return;
+		}
+	}
+
+	if (IsA(partitionValueNode, Const))
+	{
+		elog(DEBUG1, "Pushdown argument: %s",
+			 pretty_format_node_dump(nodeToString(partitionValueNode)));
```

Comment:
If we use elog the arguments always get evaluated. Should use ereport instead.

Reply:
Done